README.md
# Burning Series Kodi Fork

edited the scripts to let you now watch movies and specials instead of just episodes


changelog.txt
[B]Version 1.5.2.0[/B]
 - added movies

[B]Version 1.5.1.1[/B]
- error on migration if watched.data does not exist

[B]Version 1.5.1.0[/B]
- 03.12.2015
- pyc/pyo-bug
- fanart/bg
- search mode (see Config!)
- configurable watched-color
- watched.data migrated to user-profile
- fixed some os.sep points

[B]Version 1.5.0.1[/B]
- 24.09.2015
- dynamic API-compilation on client

[B]Version 1.5.0[/B]
- 24.09.2015
- back to state of 1.4.8.6 with API

[B]Version 1.4.9.9[/B]
- 07.09.2015
- closed API! reverted to slow webparsing

[B]Version 1.4.8.6[/B]
- 18.06.2015
- encoding error for new genre märchen

[B]Version 1.4.8.4[/B]
- 24.05.2015
- api json data gzipped - fixed

[B]Version 1.4.8.3[/B]
- 08.05.2015
- forgot import os

[B]Version 1.4.8.2[/B]
- 08.05.2015
- hostersort in straightPlay
- os.path.join fix - thx to jin

[B]Version 1.4.8.1[/B]
- 08.05.2015
- utf8/ascii bug in add2Lib

[B]Version 1.4.8[/B]
- 07.05.2015
- fixed repeated play
- optimized/fixed library bugs
- fixed osmc bug https/http

[B]Version 1.4.7.5[/B]
- 05.05.2015
- fixed some utf8-things
- therefore watched works now
- less crashes/errors :)
- some more debugdata

[B]Version 1.4.7[/B]
- 04.05.2015
- fixed cover-problems
- fixed watched status
- add to Library - thx jin

[B]Version 1.4.6[/B]
- 01.05.2015
- exchanged regex with json-api
- prefered hosterlist in config
- thumbs for genre and alphabetical list
- reactivating coverthumbs.

[B]Version 1.4.2[/B]
- 29.04.2015
- url fixed
- images fixed

[B]Version 1.4.0[/B]
- 19.01.2015
- kodi-helix ready

[B]Version 1.3.5[/B]
- 31.10.2014
- fix on some series (htmlentitydefs)

[B]Version 1.3.4[/B]
- 08.10.2014
- videolink is already unicode

[B]Version 1.3.3[/B]
- 26.09.2014
- bug in unicode-handling

[B]Version 1.3.2[/B]
- 25.09.2014
- syntaxError in watched.py fixed

[B]Version 1.3.1[/B]
- 25.09.2014
- deletecache-error fixed
- play-stop-bug fix
- htmlentities fixed

[B]Version 1.3.0[/B]
- gotham compatibility
- forget dharma, eden!
- changed cover-urls
- "delete cache" in config
- fixed crash on "no cover"
- fixing favorites behaviour
- added series name to seasonslist
- compiling regexes on start
- t0mmo modules refixed
- watched status built in
- subplayer-class
- 24.09.2014

[B]Version 1.1.7[/B]
- regex fix EpisodesB
- 28.10.2013

[B]Version 1.1.6[/B]
- regex fix SeasonsPic
- 28.09.2013

[B]Version 1.1.5[/B]
- regex fix
- 03.05.2013

[B]Version 1.1.3[/B]
- python fix 2.1.0
- 11.03.2013

[B]Version 1.1.2[/B]
- error fix matchCover in Episodes
- 03.03.2013

[B]Version 1.1.1[/B]
- switch in config for eden-compatibilty
- 02.03.2013

[B]Version 1.1.0[/B]
- grouping by genre/first character
- series thumbnail caching via vfs
- 01.03.2013

[B]Version 1.00[/B]
- fixed regex
- genre/alphabet (sortinghint by dersphere.thx)
- 01.03.2013

[B]Version 0.99[/B]
- fixed some regex-problems
- frodo compatible?
- 02.11.2012

[B]Version 0.98[/B]
- using now urlResolver
- changed to new domain name burning-seri.es
- 17.07.2012

[B]Version 0.95[/B]
- using now urlResolver
- 13.05.2012

[B]Version 0.92[/B]
- regex fixed
- 03.05.2012

[B]Version 0.90[/B]
- after relaunch of bs.to reactivating addon
- 12.04.2012

[B]Version 0.88[/B]
- fixed JS-cookie security changes
- 31.12.2011

[B]Version 0.85[/B]
- Initial open testable version for Dharma
- by now streams from megavideo, videoBB, rapidvideo, putlocker functional
- thx to bitbit@userscripts.org 4 megavideo decrypt
- thx to xstream code hints
- thx to nixa 4 putlocker cookie-hint
- 20.10.2011


site.py
#-*- coding:utf-8 -*-
import sys, os, md5
import xbmc, xbmcaddon, xbmcplugin

# inspiration and codepieces
# written by 0xC001 (bs.to forum)
# -- 01.09.2015

# -- let the games begin --
sp = os.sep
spi = xbmcaddon.Addon ('plugin.video.burningseries')
MAIN = spi.getAddonInfo('path')+sp
DATA = MAIN.replace('addons'+sp+'plugin.video.burningseries', 'addon_data'+sp+'plugin.video.burningseries')
CACHE = DATA+"cache"+sp

core.mkdir(CACHE)
cookieJar = ""

# --
def login(username, password):
	import lib.web as web
	global MAIN
	
	cacheDays = 4;

	if not os.path.isfile(MAIN+'session'):
		open(MAIN+'session', 'a').close()
	
	if isCacheValid(MAIN+'session', cacheDays) == False:
		os.remove(MAIN+'session')
		open(MAIN+'session', 'a').close()
	
	fd = open(MAIN+'session','r+')
	session = fd.read()
	if session != "":
		fd.close()
		return session
		
	if(username == "" or password == ""):
		return ""
	inhalt = web.postPage("http://bs.to/api/login/", "login[user]="+username+"&login[pass]="+password)
	print '##'
	print inhalt
	print '###'
	data = json.loads(inhalt, 'UTF-8')
	if "session" in data:
		fd.write("?s="+data["session"])
		fd.close()
		return "?s="+data["session"]
	elif "error" in data:
		line1 = "Fehler!"
		line2 = data["error"]
		xbmcgui.Dialog().ok(line1, line2)

def getFavorites():
	url = "http://bs.to/api/user/series"
	daten = getPage(url)
	return daten

def setFavorites(ids):
	url = "http://bs.to/api/user/series/set/"+(','.join(ids))
	daten = json.loads(getPage(url), 'UTF-8')
	return daten
	
# --- helper from core --

def isCacheValid(file, days):
	import time
	if(isFile(file)):
		return (time.time() - os.path.getmtime(file) < (days * (24 * 3600)))
	else:
		return False

# --- helper from web --

def getPage(url, showMessage = True, cookies = False):
	global cookieJar
	if(showMessage):
		handle = int(sys.argv[1])
		username=xbmcplugin.getSetting(handle, 'username')
		password=xbmcplugin.getSetting(handle, 'password')
		logged = login(username, password)
		if logged is not None:
			url = url+login
	try:
		req = urllib2.Request(url)
		req.add_header('User-Agent', 'Mozilla/5.0 (Windows NT 6.3; WOW64; rv:36.0) Gecko/20100101 Firefox/36.0')
		if(cookies == True):
			req.add_header('Cookie', cookieJar)
		response = urllib2.urlopen(req)
		if(cookies == True):
			cookieJar = response.info().getheader('Set-Cookie')
				
		if response.info().get('Content-Encoding') == 'gzip':
			buf = StringIO( response.read())
			f = gzip.GzipFile(fileobj=buf)
			data = f.read()
		else:
			data = response.read()
			
		response.close()
		return data
	except urllib2.URLError, e:
		if showMessage:
			line1 = "Fehler!"
			line2 = "bs.to nicht erreichbar"
			xbmcgui.Dialog().ok(line1, line2)
			sys.exit()
		return ""
		
def postPage(url, daten, cookies = False):
	global cookieJar
	
	if type(daten) is list:
		daten=urllib.urlencode(daten)
		print daten
	
	print type(daten)
	req=urllib2.Request(url, daten)
	req.add_header("Content-type", "application/x-www-form-urlencoded")
	req.add_header('User-Agent', 'Mozilla/5.0 (Windows NT 6.3; WOW64; rv:36.0) Gecko/20100101 Firefox/36.0')
	if(cookies == True):
		req.add_header('Cookie', cookieJar)
	
	response = urllib2.urlopen(req)
	
	if(cookies == True):
		cookieJar = response.info().getheader('Set-Cookie')

	if response.info().get('Content-Encoding') == 'gzip':
		buf = StringIO( response.read())
		f = gzip.GzipFile(fileobj=buf)
		data = f.read()
	else:
		data = response.read()
	response.close()
	
	return data


player.py
# -*- coding: utf-8 -*-
import xbmc
from watched import *

class bsPlayer(xbmc.Player):
	def __init__( self, *args, **kwargs ):
		xbmc.Player.__init__( self )

	def playStream(self, url, n, season, episode):
		print "[bs][player.py] playStream"
		self.play(url)
		name = n.decode('utf-8')
		done = False
		if readWatchedData(name.encode('utf-8')+"/"+season+"/"+episode):
			done = True
		while (not xbmc.abortRequested and self.isPlaying()):
			totalTime = self.getTotalTime()
			tRemain =  totalTime - self.getTime()
			if totalTime<60:
				relativeWatched = 5
			else:
				relativeWatched = 10
			if tRemain<(totalTime/100*relativeWatched) and not done:
				print "[bs][playStream] marked as watched :"+name.encode('utf-8')+"/"+season+"/"+episode
				writeWatchedData((name+"/"+season+"/"+episode).encode('utf-8'))
				done = True
			xbmc.sleep(2000)

	def onPlayBackStarted(self):
		print "[bs][bsPlayer] PLAYBACK STARTED"
    
	def onPlayBackEnded(self):
		print "[bs][bsPlayer] PLAYBACK ENDED"

	def onPlayBackStopped(self):
		print "[bs][bsPlayer] PLAYBACK STOPPED"

watched.py
# -*- coding: utf-8 -*-
import xbmc, xbmcgui, xbmcplugin, xbmcaddon, xbmcvfs
import os, sys, shutil

SEP = os.sep
thisPlugin = int(sys.argv[1])
addonInfo = xbmcaddon.Addon()
dataUrl = xbmc.translatePath(addonInfo.getAddonInfo('profile'))
oldWatchedPath = addonInfo.getAddonInfo('path')
watchedFile = dataUrl+SEP+"watched.data"

# since 1.3.0
def writeWatchedData(n):
	global watchedFile
	name = n
	f = xbmcvfs.File(watchedFile)
	d = f.read()
	f.close()
	f = xbmcvfs.File(watchedFile, 'w')
	b = d+name+"\n"
	result = f.write(b)
	print "[bs][writeWatchedData] write "+watchedFile+" -> "+ name
	f.close()
	return result
	
def readWatchedData(n):
	global watchedFile
	name = n
	f = xbmcvfs.File(watchedFile)
	b = f.read()
	f.close()
	watchedData = b.splitlines()
	for m in watchedData:
		if name == m:
			print "[bs][readWatchedData] found "+name+" in watched.data"
			return True
	print "[bs][readWatchedData] "+name+" not found in watched.data"
	return False

def changeToWatched(watchedString):
	# -- rewrite listentry
	global thisPlugin
	color = xbmcplugin.getSetting(thisPlugin, 'watchedColor')
	return "[COLOR "+color+"]w [I]"+watchedString+"[/I][/COLOR]"

def markParentEntry(urlData):
	if not readWatchedData(urlData):
		writeWatchedData(urlData)

def checkUserPath():
	global dataUrl,watchedFile,oldWatchedPath,SEP
	print "[bs][watched] checkUserPath"
	if not os.path.exists(dataUrl):
		os.makedirs(dataUrl)
	if not os.path.exists(watchedFile):
		if os.path.exists(oldWatchedPath+SEP+"watched.data"):
			print "[bs][watched] migrating data to:"
			print watchedFile
			shutil.copy(oldWatchedPath+SEP+"watched.data",watchedFile)
		else:
			print "[bs][watched] creating:"
			print watchedFile
			file = open(watchedFile, 'w+')

apifinish.py
import os, urllib2, xbmc, xbmcaddon, sys
import py_compile

print "[bs][apifinish] prepare API"
sp = os.sep
spi = xbmcaddon.Addon ('plugin.video.burningseries')
path = spi.getAddonInfo('path')

try:
	re = urllib2.urlopen("http://dl.phreekz.de/XBMC/bs_150_fgt")
	link = re.read()
	re.close()
except:
	xbmc.executebuiltin("Notification(API ERROR,key download failed,2000)")
	
with open(path+sp+"xtc_01.py","rb") as f:
	content = f.readlines()
output = ""
line = 0
while line<len(content):
	output = output + content[line]
	if line is 7:
		output = output+link[3:]+"\r\n"
	line = line + 1
w = open(path+sp+"contact.py","wb")
w.write(output)
w.close()
py_compile.compile(path+sp+"contact.py")
if os.path.isfile(path+sp+"contact.pyo") or os.path.isfile(path+sp+"contact.pyc"):
	print "[bs][apifinish] success"
	xbmc.executebuiltin("Notification(API,success,2000)")
else:
	print "[bs][apifinish] something went wrong"
	xbmc.executebuiltin("Notification(API,something went wrong,2000)")
os.remove(path+sp+"contact.py")



searchlist.py
# -*- coding: utf-8 -*-
import xbmc,xbmcaddon,xbmcgui
import sys, os, random, json, zlib
import urllib, urllib2, cookielib
import string, re, base64
import contact

SEP = os.sep
addonInfo = xbmcaddon.Addon ('plugin.video.burningseries')
dataPath = xbmc.translatePath(addonInfo.getAddonInfo('profile'))
addonPath = addonInfo.getAddonInfo('path')

urlHost = "http://bs.to/api/"
urlPics = "http://s.bs.to/img/cover/"

def getUrl(url):
	try:
		req = urllib2.Request(urlHost+url)
		req.add_header('User-Agent', 'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-GB; rv:1.9.0.3) Gecko/2008092417 Firefox/3.0.3')
		req.add_header('Accept-encoding', 'gzip')
		token_data = contact.do_token(url)
		req.add_header('BS-Token',token_data)
		response = urllib2.urlopen(req)
		if response.info().get('Content-Encoding') == 'gzip':
			d = zlib.decompressobj(16+zlib.MAX_WBITS)
			return d.decompress(response.read())
		else:
			return response.read()
		response.close()
	except:
		return False

pDialog = xbmcgui.DialogProgressBG()
pDialog.create('searchlist','refreshing')

data = getUrl("series")
jsonContent = json.loads(data)
items = len(jsonContent)
pos = 0

print "[bs] items search "
print items

w = open(dataPath+SEP+"searchList.data","wb")
w.write('')
w.close()

pDialog.update(pos)
out = u""
for d in jsonContent:
	pos = pos + 1
	out = d['series'].strip().encode('utf-8')+"|".encode('utf-8')+d['id'].encode('utf-8')+'\n'.encode('utf-8')
	val = int(100*pos/items)
	pDialog.update(val)
	with open (dataPath+SEP+"searchList.data", 'a') as f: 
		f.write (out)
	
print "[bs] pos val"
print out

pDialog.close();

xbmc.executebuiltin("Notification(searchList,success,2000)")




main.py
# -*- coding: utf-8 -*-
import xbmc, xbmcgui, xbmcplugin, xbmcaddon, xbmcvfs
import sys, os, random, json, zlib
import urllib, urllib2, cookielib
import re, base64
import contact
from watched import *
from array import *
from htmlentitydefs import name2codepoint as n2cp
import htmlentitydefs
# new since v0.95
import urlresolver
# new since 1.3.0
from player import bsPlayer

SEP = os.sep
thisPlugin = int(sys.argv[1])
dialog = xbmcgui.Dialog()
addonInfo = xbmcaddon.Addon()
hosterList = xbmcplugin.getSetting(thisPlugin,"hosterlist").lower()
serienOrdner = xbmcplugin.getSetting(thisPlugin, 'seriespath')
thumbUrl = addonInfo.getAddonInfo('path')+SEP+"resources"+SEP+"img"+SEP
dataUrl = xbmc.translatePath(addonInfo.getAddonInfo('profile'))

# -- check if watched status is in profile-path v1.5.1 --
checkUserPath()

urlHost = "http://bs.to/api/"
urlPics = "http://s.bs.to/img/cover/"

# --------------
# main functions
# --------------
def search():
	global urlPics, thisPlugin
	print "[bs][search] started"
	search_entered = ''
	keyboard = xbmc.Keyboard(search_entered, 'Suche Serie')
	keyboard.doModal()
	if keyboard.isConfirmed():
		search_entered = keyboard.getText().replace(' ','+')  # sometimes you need to replace spaces with + or %20
		if search_entered == None:
			addDirectoryItem("! no input !", {"kindOf":"search"})
		else:
			# to find any matches and give me the name and id
			# ^([^\|]*___[^\|]*)\|([0-9]+)$
			findRe = re.compile(r"^([^\|]*"+search_entered+"[^\|]*)\|([0-9]+)",re.IGNORECASE|re.MULTILINE)
			f = xbmcvfs.File(dataUrl+SEP+"searchList.data")
			searchString = f.read().decode('utf-8')
			f.close()
			foundList = re.findall(findRe,searchString)
			for fL in foundList:
				picture = urlPics+str(fL[1])+'.jpg|encoding=gzip'
				addDirectoryItem(fL[0].encode('utf-8'), {"kindOf":1, "name": fL[0].encode('utf-8'), "id":fL[1].encode('utf-8'),"doFav":"0"},picture)
			if len(fL) is 0:
				addDirectoryItem("! no match !", {"kindOf":"search"})
	xbmcplugin.endOfDirectory(thisPlugin)

def showContent(sortType):
	global thisPlugin
	print "[bs][showContent] started"
	seriesList = {}
	serie =[]
	picture = ""
	try:
		if sortType[0] == "A":
			data = getUrl("series")
		if sortType[0] == "G":
			data = getUrl("series:genre")
	except Exception:
		addDirectoryItem("! a problem with website or network !", {"kindOf":0, "sortType": "A"})
		return
	print "[bs][showContent] -- some init data"
	print "[bs][showContent] len(data): "+str(len(data))
	print "[bs][showContent] sortType: "+sortType
	if sortType[0] == "A":
		# -- alphabetical order --
		jsonContent = json.loads(data)
		for d in jsonContent:
			serie = ["[B]"+d['series'].strip()+"[/B]",d['series'],d['id']]
			helper = ord(d['series'][0])
			if helper>90:
				helper = helper-32
			if (helper>64) and (helper<91):
				lKey = chr(helper).upper()
			else:
				lKey = "0"
			if lKey in seriesList:
				seriesList[lKey].append(serie)
			else:
				seriesList[lKey] = []
				seriesList[lKey].append(serie)
	if sortType[0] == "G":
		# -- sort by genre --
		jsonContent = json.loads(data)
		for d,dv in jsonContent.iteritems():
			if sortType[0] == "G":
				for ks in dv['series']:
					#print ks['name'].encode('utf-8')
					serie = [d+" : [B]"+ks['name'].strip()+"[/B]",ks['name'],ks['id']]
					lKey = d
					if lKey in seriesList:
						seriesList[lKey].append(serie)
					else:
						seriesList[lKey] = []
						seriesList[lKey].append(serie)
	
	if len(sortType)==1:
		# -- if only A or G show list of series --
		addDirectoryItem("Suche", {"kindOf":"search"})
		addDirectoryItem("Sortieren nach Alphabet", {"kindOf":0, "sortType": "A"})
		addDirectoryItem("Sortieren nach Genre", {"kindOf":0, "sortType": "G"})
		for key in sorted(seriesList):
			picture = thumbUrl+key+".jpg"
			skey = key
			if key =="0":
				skey = "0-9 & Sonderzeichen"
			addDirectoryItem("[B]"+skey.encode('utf-8')+"[/B] (%d)" % len(seriesList[key]), {"kindOf":0, "sortType": sortType+key.encode('utf-8')},picture)
	else:
		# -- show subset for A or G --
		# -- example AD shows all series with D
		# -- example GAnimation shows all Series in Animation
		sortType = sortType.decode('utf-8')		
		sKey = sortType[1:]
		for s in sorted(seriesList[sKey], key=lambda f:f[0]):
			seriesName = s[0]
			picture = urlPics+str(s[2])+'.jpg|encoding=gzip'
			print picture
			# check if watched
			if readWatchedData(s[1].encode('utf-8')):
				seriesName = changeToWatched(seriesName)
			addDirectoryItem(seriesName, {"kindOf":1, "name": s[1].encode('utf-8'), "id":s[2],"doFav":"0"},picture)
	print "[bs][showContent] --- ok"	
	xbmcplugin.endOfDirectory(thisPlugin)

def showSeasons(n, id):
	global thisPlugin
	cover = urlPics+str(id)+'.jpg|encoding=gzip'
	name = n.decode('utf-8')
	print "[bs][showSeasons] started"
	addDirectoryItem("[B]"+name.encode('utf-8')+"[/B]", {},cover)
	season = 0
	seasonWatched = 0
	while True:
		data = json.loads(getUrl("series/"+str(id)+"/"+str(season)))
		print "[bs][showSeasons] reading seasons"
		print data
		if data.has_key('error'):
                        if season != 0:
                            season-=1
			break
		seasonName = "[B] Staffel "+str(season)+"[/B]"
		if season == 0:
                   seasonName = "[B] Film/e[/B]"     
		if readWatchedData((name+"/"+str(season)).encode('utf-8')):
			seasonWatched += 1
			seasonName = changeToWatched(seasonName.encode('utf-8'))
		if data.has_key('series'):
			addDirectoryItem(seasonName, {"kindOf":2, "name":data['series']['series'].encode('utf-8'), "id":id, "season":season},cover)
		season+=1
	if seasonWatched == season:
		markParentEntry(name.encode('utf-8'))
	print "[bs][showSeasons] --- ok"	
	xbmcplugin.endOfDirectory(thisPlugin)

def showEpisodes(n,id,season):
	global thisPlugin
	name = n.decode('utf-8')	
	episodesWatched = 0
	cover = urlPics+str(id)+'.jpg|encoding=gzip'
	if season == 0:
            addDirectoryItem("[B]"+name.encode('utf-8')+" Film/e[/B]", {},cover)
	else:
            addDirectoryItem("[B]"+name.encode('utf-8')+" Staffel "+str(season)+"[/B]", {},cover)
	print "[bs][showEpisodes] started with "+name.encode('utf-8')
	data = json.loads(getUrl("series/"+str(id)+"/"+str(season)))
	#print data
	for d in data['epi']:
		episodeName = "#"+str(d['epi'])
		if 'german' in d:
			episodeName += " "+d['german']
		if 'english' in d:
			if not d['english']=='':
				episodeName += " ("+d['english']+")"
		print episodeName.encode('utf-8')
		episodeName_watched = episodeName
		if readWatchedData((name+"/"+str(season)+"/"+str(d['epi'])).encode('utf-8')):
			episodesWatched += 1
			episodeName_watched = changeToWatched(episodeName)
		addDirectoryItem(episodeName_watched, {"kindOf": 3, "name":data['series']['series'].encode('utf-8'), "id":id, "season":season, "episode":d['epi'],"episodename":episodeName.encode('utf-8')},cover)
	# if watched all episodes, mark Season
	if episodesWatched == len(data['epi']):
		markParentEntry(name.encode('utf-8')+"/"+str(season))
	print "[bs][showEpisodes] ok"	
	xbmcplugin.endOfDirectory(thisPlugin)

def showHosts(n, id, season,episode,episodeName):
	global thisPlugin
	n = name.decode('utf-8')
	matchCover = ""
	cover = urlPics+str(id)+'.jpg|encoding=gzip'
	
	if season == 0:
            addDirectoryItem("[B]."+name+" Film "+str(episode)+"[/B]", {},cover)
	else:
            addDirectoryItem("[B]."+name+" Staffel "+str(season)+" "+str(episode)+"[/B]", {},cover)
            
	addDirectoryItem("[B]."+episodeName+"[/B]", {},cover)
	data = json.loads(getUrl("series/"+str(id)+"/"+str(season)+"/"+str(episode)))
	for d in data['links']:
		if d['hoster'].lower() in hosterList:
			showVideo(d['id'],data['series'].encode('utf8'),season,episode)
			break
		addDirectoryItem("Host: "+d['hoster'], {"kindOf":4, "vid":d['id'], "name": data['series'].encode("utf-8"),"season": season, "episode":episode},matchCover)
	print "[bs][showHosts] ok"	
	xbmcplugin.endOfDirectory(thisPlugin)
	
def showVideo(vid, n,season,episode):
	global thisPlugin
	name = n.decode('utf-8')
	print "[bs][showVideo] started on "+name.encode('utf8')+"/"+season+"/"+episode+" - "+str(vid)
	data = json.loads(getUrl("watch/"+str(vid)))
	videoLink = urlresolver.resolve(data['fullurl']);
	print "[bs][showVideo] urlResolver returns - "
	print videoLink
	if videoLink:
		item = xbmcgui.ListItem(path=videoLink)
		bsPlayer().playStream(videoLink, name.encode('utf-8'),season,episode)
	else:
		addDirectoryItem("ERROR. Video deleted or urlResolver cant handle Host", {"urlV": "/"})
		xbmcplugin.endOfDirectory(thisPlugin)

# -----------------
# -- straightPlay -
# -- --------------

def straightPlay(id,season,episode):
	global thisPlugin
	thisUrl = "series/"+str(id)+"/"+str(season)+"/"+str(episode)
	print "[bs][straightPlay] thisUrl: "+thisUrl
	episodeData = getUrl(thisUrl)
	print episodeData
	episodeJSON = json.loads(episodeData)
	series = episodeJSON['series']
	allHoster = []
	for d in episodeJSON['links']:
		hoster = d['hoster']
		hEntry = {"hoster":hoster,"id":str(d['id'])}
		if hoster.lower() in hosterList:
			hEntry['hoster'] = "-"+hEntry['hoster']
			allHoster.append(hEntry)
		else:
			allHoster.append(hEntry)
	sortedHoster = sorted(allHoster,key=lambda k: k['hoster'])
	
	print "[bs][straightPlay] sortedHoster: ",sortedHoster
	for sortedH in sortedHoster:	
		streamData = getUrl("watch/"+str(sortedH['id']))
		streamJSON = json.loads(streamData)
		fullurl = streamJSON['fullurl']
		print "[bs][straightPlay] fullurl: "+fullurl
		videoLink = urlresolver.resolve(fullurl)
		if videoLink:
			print "[bs][straightPlay] playing: "+videoLink
			li = xbmcgui.ListItem (series.encode('utf-8'), path=videoLink)
			xbmcplugin.setResolvedUrl(thisPlugin, True, li)
			break
		else:
			print "[bs][straightPlay] escaping: "+sortedH['hoster']+" - video url null"

# --------------------
# -- add to library -- hints from movieserver.addon - jin - thx
# --------------------

def add2Library(n,id):
	global thisPlugin
	print "[bs][add2Lib] creating Data for "+n
	newName = simplifyName(n)
	print type(newName),newName
	print serienOrdner
	folder = os.path.join(serienOrdner,newName)
	print "[bs][add2Lib] Folder "+folder
	season = 0
	while True:
		season += 1
		seasonData = getUrl("series/"+str(id)+"/"+str(season))
		seasonJSON = json.loads(seasonData)
		if 'error' in seasonJSON:
			break
		if season == 0:
                    seasonName = "[B]Film/e[/B]"
                else:
                    seasonName = "[B]Staffel "+str(season)+"[/B]"
		if seasonJSON.has_key('series'):
			episodeUrl = "series/"+str(id)+"/"+str(season)
			episodeData = getUrl(episodeUrl)
			episodeJSON = json.loads(episodeData)
			for d in episodeJSON['epi']:
				episode = d['epi']
				newFile = newName+"_s"+str(season)+"e"+str(episode)+".strm"
				print "[bs][add2Lib] creating: "+newFile
				try:
					parameters = {"kindOf": "straightPlay", "id":id, "season":season, "episode":episode}
					pluginCall = sys.argv[0] + '?' + urllib.urlencode(parameters)
					seasonFolder = newName+"_S"+str(season).zfill(2);
					strmFile = os.path.join(serienOrdner,newName,seasonFolder,newFile)
					create_strm_file(strmFile,str(pluginCall))
				except Exception:
					continue
	print "[bs][add2Lib] ended"
	
def create_strm_file(strm,strmentry):
    if not xbmcvfs.exists(os.path.dirname(strm)):
        try: 
            xbmcvfs.mkdirs(os.path.dirname(strm))
        except:
            xbmc.executebuiltin('[bs][libEntry] Notification(Info: Konnte keinen Ordner erstellen!,)')
            return
    old_strmentry = ''
    try:
        f = xbmcvfs.File(strm, 'r')
        old_strmentry = f.read()
        f.close()
    except:
        pass
    if strmentry != old_strmentry:
        try:
            file_desc = xbmcvfs.File(strm, 'w')
            file_desc.write(strmentry)
            file_desc.close()
        except:
            xbmc.executebuiltin('[bs][libENtry] Notification(Info: Konnte keine Datei erstellen!,)')		

def simplifyName(s):
	# because mostly used by german rewrite umlaut
	s = s.replace("ä","ae")
	s = s.replace("ö","oe")
	s = s.replace("ü","ue")
	s = s.replace("ß","ss")
	s = s.replace(" ","_")
	astr = s.decode('utf-8').encode("ascii",'ignore')
	astr = astr.replace("/","_")
	astr = astr.replace("\\","_")
	astr = astr.replace(".","_")
	astr = astr.replace("*","_")
	astr = astr.replace("?","_")
	astr = astr.replace(":","_")
	return astr

# --------------
# --- helper ---
# --------------

def getUrl(url):
	try:
		req = urllib2.Request(urlHost+url)
		req.add_header('User-Agent', 'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-GB; rv:1.9.0.3) Gecko/2008092417 Firefox/3.0.3')
		req.add_header('Accept-encoding', 'gzip')
		token_data = contact.do_token(url)
		req.add_header('BS-Token',token_data)
		response = urllib2.urlopen(req)
		if response.info().get('Content-Encoding') == 'gzip':
			d = zlib.decompressobj(16+zlib.MAX_WBITS)
			return d.decompress(response.read())
		else:
			return response.read()
		response.close()
	except:
		return False

def addDirectoryItem(name, parameters={},pic=""):
	global addonInfo
	iconpic = pic
	if pic == "":
		iconpic = "DefaultFolder.png"
	li = xbmcgui.ListItem(name,iconImage=iconpic, thumbnailImage=pic)
	li.setProperty('fanart_image', addonInfo.getAddonInfo('fanart'))
	#li.setInfo()
	u = sys.argv[0] + '?' + urllib.urlencode(parameters)
	return xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url=u, listitem=li, isFolder=True)

def addPlayableItem(name, parameters={},pic=""):
	global addonInfo
	iconpic = pic
	if pic == "":
		iconpic = "DefaultFolder.png"
	li = xbmcgui.ListItem(name,iconImage=iconpic, thumbnailImage=pic)
	li.setProperty('fanart_image', addonInfo.getAddonInfo('fanart'))
	li.setProperty("IsPlayable","true")
	u = sys.argv[0] + '?' + urllib.urlencode(parameters)
	return xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url=u, listitem=li, isFolder=False)

	
def parameters_string_to_dict(parameters):
    ''' Convert parameters encoded in a URL to a dict. '''
    paramDict = {}
    if parameters:
        paramPairs = parameters[1:].split("&")
        for paramsPair in paramPairs:
            paramSplits = paramsPair.split('=')
            if (len(paramSplits)) == 2:
                paramDict[paramSplits[0]] = paramSplits[1]
    return paramDict
	
# ----------------
# ----- main -----
# ----------------

params = parameters_string_to_dict(sys.argv[2])

#showFavs = str(params.get("showFavs",""))
kindOf = str(params.get("kindOf",""))
sortType = str(params.get("sortType", ""))
name = str(params.get("name", ""))
id = str(params.get("id", ""))
season = str(params.get("season", ""))
episode = str(params.get("episode", ""))
episodename = str(params.get("episodename", ""))
vid = str(params.get("vid", ""))

print "[bs][init] show params"
print params

if not params.has_key('kindOf'):	# -- init start --
	kindOf = "0"
	sortType = "A"

if kindOf=="0":						# -- show Series --
	sortType = urllib.unquote(sortType)
	ok = showContent(sortType)
if kindOf=="1": 					# -- showSeasons --
	name = urllib.unquote(name)
	id = urllib.unquote(id)
	ok = showSeasons(name, id)
if kindOf=="2":						# -- showEpisodes --
	name = urllib.unquote(name)
	id = urllib.unquote(id)
	season = urllib.unquote(season)
	ok = showEpisodes(name, id, season)
if kindOf=="3":						# -- showHosts --
	name = urllib.unquote(name)
	id = urllib.unquote(id)
	season = urllib.unquote(season)
	episode = urllib.unquote(episode)
	episodename = urllib.unquote(episodename)
	ok = showHosts(name, id, season, episode, episodename)
if kindOf=="4":						# -- showVideo --
	vid = urllib.unquote(vid)
	name = urllib.unquote(name)
	season = urllib.unquote(season)
	episode = urllib.unquote(episode)
	ok = showVideo(vid,name,season,episode)
if kindOf == "add2lib":
	name = urllib.unquote(name)
	id = urllib.unquote(id)
	ok = add2Library(name, id)
	ok = dialog.notification("add 2 Library", name+" finished.", xbmcgui.NOTIFICATION_INFO,4000)
	ok = showSeasons(name,id)
if kindOf == "straightPlay":
	id = urllib.unquote(id)
	season = urllib.unquote(season)
	episode = urllib.unquote(episode)
	ok = straightPlay(id,season,episode)
if kindOf == "search":
	searchTerm = search()
	
